### ISSUE TITLE
Adding is_filtered_index for SQL Server dialect

### ISSUE THREAD

**Issue by pmantica1** [sqlalchemy/sqlalchemy#4966]
I noticed that SQLAlchemy does not reflect the filter definition of SQL Server indexes. I see how this makes sense since you expect the filter definition to be a SQLAlchemy construct instead of a SQL string.
 
https://docs.sqlalchemy.org/en/13/dialects/mssql.html#filtered-indexes

However, it is not possible to determine if a unique index implies uniqueness in a subset of the table or the entire table without this information.  Would be possible to add an `is_filtered_index` dialect-specific kwarg for MSSQL Server or something along these lines? I would be happy to implement it and submit a PR.

**Comment by zzzeek** [comment]
these can be reflected if you can get the SQL text for them, they would be fulfilled using the text() construct.   the mssql_where then goes in to the "dialect_options" entry of the index entry.

**Comment by zzzeek** [comment]
then you can tell if it's a filtered index by the presence of that existing argument.   we need this for PostgreSQL also.

**Comment by zzzeek** [comment]
PR needs to have tests also and for this you can likely go in https://github.com/sqlalchemy/sqlalchemy/blob/master/test/dialect/mssql/test_reflection.py#L286 with the other SQL Server specific index reflection tests.

**Comment by pmantica1** [comment]
> these can be reflected if you can get the SQL text for them, they would be fulfilled using the text() construct. the mssql_where then goes in to the "dialect_options" entry of the index entry.

When you say "text construct" do that mean that SQLAlchemy has the ability to convert SQL text of the SQL Server dialect to a SQLAlchemy python object?

**Comment by zzzeek** [comment]
> > these can be reflected if you can get the SQL text for them, they would be fulfilled using the text() construct. the mssql_where then goes in to the "dialect_options" entry of the index entry.
> 
> When you say "text construct" do that mean that SQLAlchemy has the ability to convert SQL text of the SQL Server dialect to a SQLAlchemy python object?

it doesn't "convert" it, it just wraps it in the appropriate representation: https://docs.sqlalchemy.org/en/13/core/sqlelement.html?highlight=text#sqlalchemy.sql.expression.text

if you for example reflect a CHECK constraint, you get a CheckConstraint back which has the SQL text inside of a text() object.

**Comment by RamonWill** [comment]
hi @zzzeek ,

I'd be happy to look into this. I still need to do more testing on my solution as well as implement it for PostgreSQL but I am thinking of doing something like the below (just wanted to double check with you to make sure I'm on the right track):
```diff
# lib\sqlalchemy\dialects\mssql\base.py  line 2754
        for row in rp.mappings():
            if row["index_id"] in indexes:
                indexes[row["index_id"]]["column_names"].append(row["name"])

+        rp_filtered_indexes = connection.execution_options(
+            future_results=True
+            ).execute(
+            sql.text(
+                "SELECT ind.index_id, ind.Name, "
+                "ind.type_desc, ind.filter_definition "
+                "FROM sys.indexes as ind "
+                "INNER JOIN sys.tables as tab ON "
+                "tab.object_id = ind.object_id "
+                "INNER JOIN sys.schemas as sch ON "
+                "sch.schema_id = tab.schema_id "
+                "WHERE tab.name =:tabname "
+                "AND sch.name =:schname "
+                "AND ind.has_filter = 1"
+            )
+            .bindparams(
+                sql.bindparam("tabname", tablename, ischema.CoerceUnicode()),
+                sql.bindparam("schname", owner, ischema.CoerceUnicode()),
+            ))
+        for row in rp_filtered_indexes.mappings():
+            if row["index_id"] in indexes:
+                indexes[row["index_id"]]["is_filtered_index"] = True

+        return list(indexes.values())

```

The below is the test case i set up locally for it:
```
from sqlalchemy import Column, Integer, String, create_engine, Index, schema, inspect
from sqlalchemy.orm import base, sessionmaker
from sqlalchemy.sql import text
from sqlalchemy.ext.declarative import declarative_base
engine  = create_engine("mssql+pyodbc://server/database?Trusted_Connection=yes", echo=True)
Base = declarative_base()
class Users(Base):
    __tablename__ = "users"
    id = Column(Integer, primary_key=True)
    name = Column(String(40))
    age = Column(Integer)
    __table_args__ = (Index("idx_age", age, mssql_where= age >= 18),)

Base.metadata.create_all(engine)
insp = inspect(engine)
all_indexes = insp.get_indexes("users")  # [{'name': 'idx_age', 'unique': False, 'column_names': ['age'], 'is_filtered_index': True}]
print(all_indexes[0]["is_filtered_index"])  # prints True
```

**Comment by zzzeek** [comment]
hey there.

Can we just augment the first query at the top of def get_indexes() with the extra columns?   that way we can just use the query we already have.

**Comment by zzzeek** [comment]
then it also needs "filter_definitiion" I would assume, to come out as "mssql_where" in the dialect_options dictionary.

**Comment by sqla-tester** [comment]
**RamonWill** has proposed a fix for this issue in the **master** branch:

**fixes #4966** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2142

**Comment by sqla-tester** [comment]
**RamonWill** has proposed a fix for this issue in the **master** branch:

**fixes #4966** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2142
